### ISSUE TITLE
Threads UX improvements ideas

### ISSUE THREAD

**Issue by hiiamboris** [Nheko-Reborn/nheko#1384]
### The Problem

Coming from https://github.com/Nheko-Reborn/nheko/issues/1273

You've built the only usable desktop client so far, and one that respects OS theme (many thanks for that! saves me the trouble of implementing it myself :) But that also means you can't get away with "that's how I like it" answer, for now your userbase is big whether you like it or not ;) Plus even when you're not starting threads, you're bound to look at them when someone else starts.

I'm not going to promote the moronic implementation of Element (which is the same as it was in Gitter), where you have to sometimes scroll up 10+ screens to find where someone left an answer. Threads UI/UX is a historical PITA, and Nheko implementation is probably the best one so far. But I would like it to be better still. 

First, look at the UX shortcomings with me please:

1. I can't even tell the threads apart by their yarn color:

   ![](https://i.gyazo.com/45673dc41be3b4e73ab8652d40e5cffe.png)

   Last two messages belong to different threads but visually I don't see that, since the colors are too close.

2. It makes it easy to do the wrong thing: write the answer into the global conversation.

   Because the yarn icon is so tiny and is located at the far right side (not where eyes of the reader are likely to go), it looks like it belongs to the main conversation, and facilitates answers into it. But others will see this message as out of place and will be confused for sure. 

3. There's no connection between the thread topic and the answer, which complicates reasoning about threads topology.

### The Solution

What I suggest is to rearrange UI items a bit and explain why:

![](https://i.gyazo.com/20cc97d1f357eccd1150bb52d5927f6c.png)

There can of course be many variants of the solutions to above problems:
- Instead of an arrow, initial message could be differentiated from the new messages by:
  - painting in the yarn color (but that would make the log look like a rainbow - not great)
  - painting it in simply less pronounced color (closer to background - but may become less readable in daylight)
  - changing the background color (like zebra in tables)
  - using smaller font
  - it can itself be shortened
  
- While initial message should be shortened/ellipsized when it's more than one line, new messages may or may not appear in full.

  Personally I would show only the first line of new message: this way I will immediately be able to read answers like "OK" but will have to open the thread to read any in-depth answers, and at the same it would keep the main log not so cluttered by threads.
  
- To show thread starting message or previous message in the thread?

  Starting message will make easier to identify threads IMO, but there's no one size fits all answer.

- Your own ideas :)
  
What do you think @deepbluev7 ? It should be a piece of cake for you to implement, so please consider it :)

### Happens in the latest version

- [X] Yes, this feature is still missing.

**Comment by LorenDB** [comment]
> But that also means you can't get away with "that's how I like it" answer, for now your userbase is big whether you like it or not ;)
I feel obligated to point out that technically, the maintainers can make whatever changes they want. Nobody is under any obligation here. But we are certainly happy to try to implement good feedback in this case :)

With that being said, I'm not sure that I think this an entirely perfect approach. I'd personally like to stay away from repeating the first message every time the thread is replied to.

**Comment by hiiamboris** [comment]
> I'd personally like to stay away from repeating the first message every time the thread is replied to.

Yes, it's a concern indeed :) I clearly wouldn't want to see the *whole* message repeated every time, that's why I offer only showing the first line and a few ways to make it less distracting. Another option perhaps is to show the original message on hover (in a popup), and while this is not keyboard-friendly, current reply/react/start-thread UX isn't either, so worth considering. The art here is to find the middle ground between two extremes: annoying message display, and annoying lack of it ;)

**Comment by Cadair** [comment]
I agree that the thread indicator could be moved to be more obvious, I also sometimes miss it. I can also see that only using colour of a small icon to indicate which thread is which could be an accessibility concern.

I am a hard :-1: on repeating any part of the previous messages in the thread display though. I personally very much appreciate that it's not there as it makes it clearly distinct from replies and doesn't clutter the UI. I think there are many other options (some like you list) which could make it clear that a message is in a thread, and then if your room list is cluttered you can click through to just the single thread view to see context.

**Comment by hiiamboris** [comment]
> Additionally we could start showing the thread indicator for the original message as well, even if that is a bit more difficult. then you might have one solid colored line for the related messages.

Sounds good. Since it's gonna be small, we'll need to be able to click on the message to open a thread view. Although I notice that normal replies are currently drawn with solid colored box on the left. How are we going to discern them?

**Comment by deepbluev7** [comment]
> how are we going to discern them?

Well, for replies the colored bar is immediately followed by text inside the same message without such a bar. For threads it would be along the whole message. But yes, it might be a bit hard to tell.

**Comment by hiiamboris** [comment]
Another note: once threaded messages start to prevail, it becomes harder and harder to visually track normal room talk.

**Comment by quazgar** [comment]
### Front marker ###

> I'd probably instead just move the thread indicator as a colored bar to the left of the message, similar to how replies are rendered, but in the thread color and without a background. That should make it harder to miss.

Putting some marker in front, whether a bar, a thread logo or both, sound really good UI wise.

### Highlighting ###

I could also imagine that hovering over one message of a thread highlights all the other messages of the same thread, maybe by setting their background to a light tint of the thread's color, or flashing up the color bar.

**Comment by Cadair** [comment]
> Maybe there could be a colored bar long the whole window when one is inside a thread.

Some kind of coloured overline or border around the text input box could certainly work :+1:

**Comment by deepbluev7** [comment]
FTR, these are the current thread markers:

![Screenshot_20230531_230607](https://github.com/Nheko-Reborn/nheko/assets/16312324/c4c704a5-23ef-45cf-ad96-e7b6b96f64f5)

**Comment by pizdjuk** [comment]
> Another note: once threaded messages start to prevail, it becomes harder and harder to visually track normal room talk.

I really and honestly second this. Right now I found the manner like Element solved threading not the best, but superior that nheko have right now. So, why it is superior?

1. The thread-messages are not mixed into the main conversation and between other threads. You can simply follow on the main conversation. The timeline on the messages is secondary in respect of the sense of the messages. So it is better to have the messages ordered respect to threads and not respect to timeline.

2. The are list of all threads in a conversations. It is a pretty useful method to maintain different subjects in a room, or to pin some stuff for later responses.

Nheko could:

Option1: simply copy the manner Element manages threads. It means blend out thread messages from the main conversation and implement a list of threads somewhere with possibility to focus on any thread.

Option2: implement the tree view of threads, but anyway there is need of a list of all threads. But as the tree could be too big, as threads can grow, I would appreciate the Option1 over Option2.

As for now I don't have any better idea, then simply copy the methodology of Element. To defend this, I can refer on Telegram, that manages threads in the same manner and it seems to proved as useful method.

**Comment by hiiamboris** [comment]
Option 3 could be to have an *easily* accessible per-room switch that filters threads out of the main convo, but leave the Nheko UX as is. It is cool, until it faces scaling problems.

Also, discord solved it much better than element. In discord you see all thread headers without moving anywhere into another window, and you don't have to look at two chats at the same time (which is total idiocy).
